iq/iq.go

```diff
@@ -0,0 +1,212 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package iq has definitions and functions for processing golang purls with Nexus IQ Server
+package iq
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/package-url/packageurl-go"
+	"github.com/sonatype-nexus-community/nancy/buildversion"
+	"github.com/sonatype-nexus-community/nancy/configuration"
+	"github.com/sonatype-nexus-community/nancy/customerrors"
+	"github.com/sonatype-nexus-community/nancy/cyclonedx"
+	"github.com/sonatype-nexus-community/nancy/ossindex"
+	"github.com/sonatype-nexus-community/nancy/types"
+)
+
+const internalApplicationIDURL = "/api/v2/applications?publicId="
+
+const thirdPartyAPILeft = "/api/v2/scan/applications/"
+
+const thirdPartyAPIRight = "/sources/nancy?stageId="
+
+const (
+	pollInterval = 1 * time.Second
+)
+
+var localConfig configuration.IqConfiguration
+
+// Internal types for use by this package, don't need to expose them
+type applicationResponse struct {
+	Applications []application `json:"applications"`
+}
+
+type application struct {
+	ID string `json:"id"`
+}
+
+type thirdPartyAPIResult struct {
+	StatusURL string `json:"statusUrl"`
+}
+
+var statusURLResp types.StatusURLResult
+
+func getPurls(purls []string) (result []packageurl.PackageURL) {
+	for _, v := range purls {
+		purl, _ := packageurl.FromString(v)
+		result = append(result, purl)
+	}
+
+	return
+}
+
+// AuditPackages accepts a slice of purls, public application ID, and configuration, and will submit these to
+// Nexus IQ Server for audit, and return a struct of StatusURLResult
+func AuditPackages(purls []string, applicationID string, config configuration.IqConfiguration) types.StatusURLResult {
+	localConfig = config
+
+	if localConfig.User == "admin" && localConfig.Token == "admin123" {
+		warnUserOfBadLifeChoices()
+	}
+
+	internalID := getInternalApplicationID(applicationID)
+	resultsFromOssIndex, err := ossindex.AuditPackages(purls)
+	customerrors.Check(err, "There was an issue auditing packages using OSS Index")
+
+	sbom := cyclonedx.ProcessPurlsIntoSBOM(resultsFromOssIndex)
+	statusURL := submitToThirdPartyAPI(sbom, internalID)
+
+	finished := make(chan bool)
+
+	statusURLResp = types.StatusURLResult{}
+
+	go func() {
+		for {
+			select {
+			case <-finished:
+				return
+			default:
+				pollIQServer(fmt.Sprintf("%s/%s", localConfig.Server, statusURL), finished)
+				time.Sleep(pollInterval)
```

Comment:
So if i understand this code path correctly we will continually just poll IQ Server indefinitely?? Should we at least stop after `n` number of retries at least??

Reply:
Did a bit in: 180312c